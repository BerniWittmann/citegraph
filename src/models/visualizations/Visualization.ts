import Transformer from '@/transformers/Transformer'

export interface TimePeriod {
  from: number
  to: number
}

export interface VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods?: Array<TimePeriod>
}

export interface VisualizationBaseFieldsDTO {
  id?: string;
  progress?: number;
  name: string;
  key: string;
  // eslint-disable-next-line camelcase
  time_periods?: Array<TimePeriod>
}

export default abstract class Visualization implements VisualizationBaseFields {
  id?: string;
  progress?: number;
  name: string;
  timePeriods: Array<TimePeriod>
  abstract parameters: Object;
  abstract data?: Object;
  abstract readonly key: string;
  static readonly key: string;
  static readonly title: string;
  static readonly imageUrl: string;
  static readonly description: string;
  static readonly longDescription: string;
  static readonly transformer: typeof Transformer;

  protected constructor ({ id, progress, name, timePeriods }: VisualizationBaseFields) {
    this.id = id
    this.progress = progress
    this.name = name
    this.timePeriods = timePeriods || []
  }
}
