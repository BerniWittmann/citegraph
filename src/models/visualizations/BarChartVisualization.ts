import Visualization, { VisualizationBaseFields } from './Visualization'

export interface BarChartVisualizationFields {}

export default class BarChartVisualization extends Visualization implements BarChartVisualizationFields {
  static key: string = 'barchart'
  key: string = 'barchart'
  static title: string = 'visualizations.information.bar_chart.title'
  static description: string = 'visualizations.information.bar_chart.description'
  static longDescription: string = 'visualizations.information.bar_chart.long_description'
  static imageUrl: string = 'visualization_images/bar_chart.png'

  constructor ({ id, name, progress }: VisualizationBaseFields & BarChartVisualizationFields) {
    super({ id, name, progress })
  }
}
