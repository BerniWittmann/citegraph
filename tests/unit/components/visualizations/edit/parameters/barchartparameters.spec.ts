import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../../setupPlugins'

import Visualization from '@/models/visualizations/Visualization'
import BarChartVisualization from '@/models/visualizations/BarChartVisualization'
import BarChartParameters from '@/components/visualizations/edit/parameters/BarChartParameters.vue'

jest.useFakeTimers()

describe('components/visualizations/edit/parameters/BarChartParameters.vue', () => {
  const getWrapper = (vis: Visualization = new BarChartVisualization({ name: 'Test' })) => {
    return shallowMount(BarChartParameters, {
      i18n,
      mocks: {
        $t: (key: string) => key
      },
      propsData: {
        visualization: vis
      }
    })
  }

  it('renders', () => {
    const wrapper = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can submit the form', () => {
    const wrapper = getWrapper()
    const btn = wrapper.findAll('v-btn-stub').at(1)
    expect(btn.exists()).toBeTruthy()
    const form = wrapper.find('v-form-stub')
    expect(form.exists()).toBeTruthy()
    // @ts-ignore
    wrapper.vm.valid = true
    expect(wrapper.emitted('submit')).toBeFalsy()

    form.vm.$emit('submit', { preventDefault: jest.fn() })

    expect(wrapper.emitted('submit')).toBeTruthy()
  })

  it('can go back', () => {
    const wrapper = getWrapper()
    const btn = wrapper.find('v-btn-stub')
    expect(btn.exists()).toBeTruthy()
    expect(wrapper.emitted('back')).toBeFalsy()
    btn.vm.$emit('click')

    expect(wrapper.emitted('back')).toBeTruthy()
  })

  it('prevents the form submit if it is not valid', () => {
    const wrapper = getWrapper()
    const btn = wrapper.findAll('v-btn-stub').at(1)
    expect(btn.exists()).toBeTruthy()
    const form = wrapper.find('v-form-stub')
    expect(form.exists()).toBeTruthy()
    // @ts-ignore
    wrapper.vm.valid = false
    expect(wrapper.emitted('submit')).toBeFalsy()

    form.vm.$emit('submit', { preventDefault: jest.fn() })

    expect(wrapper.emitted('submit')).toBeFalsy()
  })
})
