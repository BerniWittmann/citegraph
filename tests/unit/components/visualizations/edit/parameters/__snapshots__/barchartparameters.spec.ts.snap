// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/visualizations/edit/parameters/BarChartParameters.vue renders 1`] = `
<v-container-stub tag="div">
  <v-form-stub class="mt-12">
    <v-row-stub tag="div" justify="center">
      <v-col-stub lg="6" tag="div">
        <v-slider-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="5" backgroundcolor="" label="visualizations.add.parameters.barchart.prune_least_cited" loaderheight="2" max="100" min="0" step="1" thumblabel="always" thumbsize="32" ticklabels="" ticksize="2"></v-slider-stub>
      </v-col-stub>
    </v-row-stub>
    <v-row-stub tag="div" justify="space-between" class="my-10">
      <v-col-stub cols="auto" tag="div">
        <v-btn-stub color="warning" tag="button" activeclass="" large="true" rounded="true" type="button">
          <v-icon-stub>mdi-menu-left</v-icon-stub>
          visualizations.add.parameters.back
        </v-btn-stub>
      </v-col-stub>
      <v-col-stub cols="auto" tag="div">
        <v-btn-stub color="success" tag="button" activeclass="" disabled="true" large="true" rounded="true" type="submit">visualizations.add.parameters.submit
        </v-btn-stub>
      </v-col-stub>
    </v-row-stub>
  </v-form-stub>
</v-container-stub>
`;
